include/rmm/device_buffer.hpp

```diff
@@ -0,0 +1,325 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <rmm/mr/default_memory_resource.hpp>
+#include <rmm/mr/device_memory_resource.hpp>
+
+#include <cuda_runtime_api.h>
+#include <cassert>
+
+namespace rmm {
+/**---------------------------------------------------------------------------*
+ * @file device_buffer.hpp
+ * @brief RAII construct for device memory allocation
+ *
+ * This class allocates untyped and *uninitialized* device memory using a
+ * `device_memory_resource`. If not explicitly specified, the memory resource
+ * returned from `get_default_resource()` is used.
+ *
+ * @note Unlike `std::vector` or `thrust::device_vector`, the device memory
+ * allocated by a `device_buffer` is unitialized. Therefore, it is undefined
+ * behavior to read the contents of `data()` before first initializing it.
+ *
+ * Examples:
+ * ```
+ * //Allocates at least 100 bytes of device memory using the default memory
+ * //resource
+ * device_buffer buff(100);
+ *
+ * // allocates at least 100 bytes using the custom memory resource
+ * custom_memory_resource mr;
+ * device_buffer custom_buff(100, &mr);
+ *
+ * // deep copies `buff` into a new device buffer
+ * device_buffer buff_copy(buff);
+ *
+ * // shallow copies `buff` into a new device_buffer, `buff` is now empty
+ * device_buffer buff_move(std::move(buff));
+ *
+ * // Default construction. Buffer is empty
+ * device_buffer buff_default{};
+ *
+ * // If the requested sized is larger than the current
+ * // size, resizes allocation to the new size and deep copies any previous
+ * // contents. Otherwise, simply updates the value of `size()` to the newly
+ * // requested size without any allocations or copies
+ * buff_default.resize(100);
+ *
+ * // Not allowed. Copy assignment is deleted
```

Comment:
Can you explain why they are deleted?

Reply:
See comments on the copy/move assignment operators.